README.md
# vex4pi

[![NPM Version](https://img.shields.io/npm/v/vex4pi.svg)](https://www.npmjs.com/package/vex4pi)

A JavaScript library allowing the Raspberry Pi to control VEX motors via the [Adafruit Servo/PWM HAT](https://www.adafruit.com/product/2327)!

# Disclaimer
* This library is intended to work only for the Raspberry Pi, and has **NOT** been tested with other hardware platforms.
* This library requires a [Adafruit Servo/PWM HAT](https://www.adafruit.com/product/2327), and will **NOT** work on direct GPIO.
* This library only supports VEX Motors at of this release, and has not been tested with other VEX hardware, such as the VEX Servos.

# Features
* Fail-safe error handling (will reset all pin speed values to 0 and close the connection to the PWM/Servo HAT before throwing an error)
* Full access to all 16 pin outputs on the PWM/Servo HAT
* Easy to use speed values compared to RobotC code (range from -100 to 100)

# Code Example
Let's say we made a simple robot with two motors, one motor controlling the right side of the wheels, and one motor controlling the left side. This code would make the robot go foward at full speed for 5 seconds before stopping. In this example, pin 0 refers to the right motor and pin 1 to the left motor.

```javascript
const { Vex4Pi } = require('vex4pi');
const { sleep } = require('sleep');

let robot = new Vex4Pi();

// initialize the pins on the robot
robot.init();

// make the robot go full speed ahead
robot.setSpeed(0, 100);
robot.setSpeed(1, -100);

// wait for 5 seconds
sleep(5);

// stop the robot
robot.setSpeed(0, 0);
robot.setSpeed(1, 0);

// free up the pins
robot.deinit();
```

# API Reference
```javascript
/*
 * Initializes the connection to the PWM/Servo HAT at address `0x40` at `/dev/i2c-1`, and
 * resets all of the pin speed values to 0.
 */
Vex4Pi.init();

/*
 * Resets all of the pin speed values to 0 before gracefully closing the connection with
 * the PWM/Servo HAT.
 */
Vex4Pi.deinit();

/*
 * Sets the `speed` of a motor at the designated `pin` value. Note that the accepted speed
 * values are between -100 and 100, and any value outside that range will be confined into
 * that range (ex. -150 => -100 or 420 => 100). If the init() function has not been called
 * before, or the pin value is not an integer/out of range, this module will throw an error.
 */
Vex4Pi.setSpeed(Integer pin, Integer speed);
```

CHANGELOG.md
## Changelog

### 1.x (Next)
* Your commit goes here

### 1.0.1 (July/9/2019)
* Updated dependencies to resolve moderate DDoS security vulnerabilities.

### 1.0.0 (Oct/9/2018)

* First official release - [@tejashah88](https://github.com/tejashah88)
  * If you were using this library prior to 1.0, please see the [UPGRADING](UPGRADING.md) document for migration instructions
* Refactored code to follow ES6 OOP style - [@tejashah88](https://github.com/tejashah88)
* Dropped support for node <6.0 - [@tejashah88](https://github.com/tejashah88)

vex4pi.js
const i2c = require('i2c-bus');
let bus;            // the bridge to directly control the vex motors
let ready = false;  // used to see if the PWM HAT is ready to send commands
let vex4pi = {};

function broadcastError(msg) {
  // always deinitialize the bus before throwing an error
  vex4pi.deinit();
  throw new Error(msg);
}

function resetAllPins() {
  for (let pin = 0; pin < 16; pin++)
    vex4pi.setSpeed(pin, 0);
}

vex4pi.init = function init() {
  bus = i2c.openSync(1);
  bus.writeByteSync(0x40,0x00,0x10); // sleep mode
  bus.writeByteSync(0x40,0xFE,0x85); // set to ~49.5Hz
  bus.writeByteSync(0x40,0x00,0x00); // wake up
  bus.writeByteSync(0x40,0x01,0x04); // output enable

  // reset the pins to avoid surprises from the motors
  resetAllPins();

  ready = true; // mark the 'ready' flag to true to allow access
};

vex4pi.deinit = function deinit() {
  // we mark the 'ready' flag false first, in order to prevent accidental access during the procedure
  ready = false;

  // reset the motor speed to neutral before closing the connection with the bus
  resetAllPins();

  // close the connection with the bus
  bus.closeSync();
}

vex4pi.setSpeed = function(pin, speed) {
  if (!ready) {
    broadcastError("I2C bus connection is not initialized with vex4pi.init()");
  } else if (pin < 0 || pin > 15) { // check if the pin value given is within range
    broadcastError("Pin value " + pin + " is out of range.");
  } else {
    // check if the speed value given is within range
    if (speed < -100)
      speed = -100;
    if (speed > 100)
      speed = 100;
  }

  //initial configuration
  bus.writeByteSync(0x40, 0x06 + (this.pin * 0x04), 0x00);
  bus.writeByteSync(0x40, 0x07 + (this.pin * 0x04), 0x00);

  // perform the calculations to convert the percentage into a PWM-ready number
  // 0x130 = neutral (speed = 0)
  speed += 0x130;
  let speedHSB = parseInt(speed / 0x100);
  let speedLSB = speed % 0x100;

  // set the actual speed
  bus.writeByteSync(0x40, 0x08 + (this.pin * 0x04), speedLSB); // low part of data
  bus.writeByteSync(0x40, 0x09 + (this.pin * 0x04), speedHSB); // high part of data
}

module.exports = vex4pi;

src/sync-api.js
'use strict';

const i2c = require('i2c-bus');
const Vex4PiError = require('./vex-error');

class Vex4Pi {
  constructor() {
    this.ready = false; // used to see if the PWM HAT is ready to send commands
  }

  resetAllPins() {
    for (let pin = 0; pin < 16; pin++)
      this.setSpeed(pin, 0);
  }

  init() {
    this.bus = i2c.openSync(1); // the bridge to directly control the vex motors
    this.bus.writeByteSync(0x40, 0x00, 0x10); // sleep mode
    this.bus.writeByteSync(0x40, 0xFE, 0x85); // set to ~49.5Hz
    this.bus.writeByteSync(0x40, 0x00, 0x00); // wake up
    this.bus.writeByteSync(0x40, 0x01, 0x04); // output enable

    // reset the pins to avoid surprises from the motors
    this.resetAllPins();

    this.ready = true; // mark the 'ready' flag to true to allow access
  }

  deinit() {
    // we mark the 'ready' flag false first, in order to prevent accidental access during the procedure
    this.ready = false;

    // reset the motor speed to neutral before closing the connection with the bus
    this.resetAllPins();

    // close the connection with the bus
    this.bus.closeSync();
  }

  setSpeed(pin, speed) {
    if (!this.ready) {
      this.deinit();
      throw new Vex4PiError('I2C bus connection is not initialized with this.init()');
    }

    // check if the pin value given is within range
    if (pin < 0 || pin > 15) {
      this.deinit();
      throw new Vex4PiError(`Pin value ${pin} is out of range.`);
    }

    // check if the speed value given is within range
    if (speed < -100) speed = -100;
    if (speed > 100) speed = 100;

    //initial configuration
    this.bus.writeByteSync(0x40, 0x06 + (this.pin * 0x04), 0x00);
    this.bus.writeByteSync(0x40, 0x07 + (this.pin * 0x04), 0x00);

    // perform the calculations to convert the percentage into a PWM-ready number
    // 0x130 = neutral (speed = 0)
    speed += 0x130;
    const speedHSB = parseInt(speed / 0x100);
    const speedLSB = speed % 0x100;

    // set the actual speed
    this.bus.writeByteSync(0x40, 0x08 + (this.pin * 0x04), speedLSB); // low part of data
    this.bus.writeByteSync(0x40, 0x09 + (this.pin * 0x04), speedHSB); // high part of data
  }
}

module.exports = Vex4Pi;